fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryRecursiveUnionPlan.java

```diff
@@ -0,0 +1,498 @@
+        return AvailableFields.ALL_FIELDS;
+    }
+
+    @Nonnull
+    @Override
+    public PRecursiveUnionQueryPlan toProto(@Nonnull final PlanSerializationContext serializationContext) {
+        final var builder = PRecursiveUnionQueryPlan.newBuilder()
+                .setInitialStateQuantifier(initialStateQuantifier.toProto(serializationContext))
+                .setRecursiveStateQuantifier(recursiveStateQuantifier.toProto(serializationContext))
+                .setInitialTempTableAlias(tempTableScanAlias.getId())
+                .setRecursiveTempTableAlias(tempTableInsertAlias.getId());
+        return builder.build();
+    }
+
+    @Nonnull
+    public static RecordQueryRecursiveUnionPlan fromProto(@Nonnull final PlanSerializationContext serializationContext,
+                                                          @Nonnull final PRecursiveUnionQueryPlan recordQueryUnorderedDistinctPlanProto) {
+        final var initialStateQuantifier = Quantifier.Physical.fromProto(serializationContext, recordQueryUnorderedDistinctPlanProto.getInitialStateQuantifier());
+        final var recursiveStateQuantifier = Quantifier.Physical.fromProto(serializationContext, recordQueryUnorderedDistinctPlanProto.getRecursiveStateQuantifier());
+        return new RecordQueryRecursiveUnionPlan(initialStateQuantifier, recursiveStateQuantifier,
+                CorrelationIdentifier.of(recordQueryUnorderedDistinctPlanProto.getInitialTempTableAlias()),
+                CorrelationIdentifier.of(recordQueryUnorderedDistinctPlanProto.getRecursiveTempTableAlias()));
+    }
+
+    @Nonnull
+    @Override
+    public PRecordQueryPlan toRecordQueryPlanProto(@Nonnull final PlanSerializationContext serializationContext) {
+        return PRecordQueryPlan.newBuilder().setRecursiveUnionQueryPlan(toProto(serializationContext)).build();
+    }
+
+    @Nonnull
+    @Override
+    public PlannerGraph rewritePlannerGraph(@Nonnull final List<? extends PlannerGraph> childGraphs) {
+        return PlannerGraph.fromNodeAndChildGraphs(
+                new PlannerGraph.OperatorNodeWithInfo(this, NodeInfo.RECURSIVE_UNION_OPERATOR),
+                childGraphs);
+    }
+
+    @Override
+    public boolean isReverse() {
+        return false;
+    }
+
+    @Override
+    public void logPlanStructure(final StoreTimer timer) {
+        timer.increment(FDBStoreTimer.Counts.PLAN_RECURSIVE_UNION);
+        getInitialStatePlan().logPlanStructure(timer);
+        getRecursiveStatePlan().logPlanStructure(timer);
+    }
+
+    @Override
+    public int getComplexity() {
+        return computeComplexity.get();
+    }
+
+    private int computeComplexity() {
+        // the complexity is calculated statically, it does not convoy the recursive nature of the
+        // actual execution of this operator, which for the most part can not be statically determined anyway.
+        return getChildren().stream().map(QueryPlan::getComplexity).reduce(1, Integer::sum);
```

Comment:
Add at least `1` on top of it. You are pretty complex here. :-)

Reply:
Oops! fixed